[WIP] Update to the recent embedded ecosystem

eupn opened the pull request:
Resolves #15.

Updated this crate to recent embedded ecosystem, including:
* Building on stable rust
* Last versions of `cortex-m-*` crates
* Updated `gcc` to newer `cc` in `build.rs`
* Updated `.cargo/config` file
* Updated examples to newer file structure (`#[entry]`)

For now, only `blinky` example is working and tested, for others I have to go through some linking issues.

eupn committed:
Update project to stable rust and recent cortex-m

eupn committed:
Update examples and SETUP.md

eupn committed:
rustfmt

jamesmunns commented on build.rs:
```diff
@@ -42,7 +42,7 @@ fn main() {
     }
 
-    process_linker_file(&outdir);
+    //process_linker_file(&outdir);
```
Is this not needed anymore? If so, could you please remove it?

eupn replied:
I thought that original linker files are no longer required to build, however, for examples besides blinky there is some linkage issues that I will figure soon hopefully. If it turns out that original linker files were required, I'll uncomment this.

jamesmunns commented:
Overall looks wonderful, thank you @eupn!

Could you please confirm that the other examples still work, particularly `ble_app_template.rs`? If that is good to go, I'm happy to merge this!

eupn commented:
As I mentioned earlier, there are some linker issues with logging and BLE examples, as soon as I figure them out it is good to go.

eupn committed:
Fix some linkage errors, but not all

eupn committed:
Make only one linker error is left

mcvicuna commented:
Is this a link issue due to having libc included by accident or is libc included on purpose?

dwaite commented:
@eupn I played around with this a bit today. To device.x, I needed to add 
```
GROUP(AS_NEEDED(-lgcc -lc -lnosys))
```

and to also define `end` (at least for `sbrk`). I did the following there, but my linker script fu is poor so I'm not sure this was correct.

```
PROVIDE(end = __sheap)
```

I went ahead and removed the start-group/end-group and library imports from `.cargo/config` as well as updating crates to their newest version.

My 52840DK board would step through in gdb but not flash lights for blinky, presumably because of the board difference. Trying to get 15.x SDK running is more than I want to bite off at this point in time. I didn't test the actual bluetooth example, just that it built.

eupn commented:
@dwaite Hello and thank you for your time and effort! Could you please make a PR with those changes against my branch?

dwaite commented:
Sure - I was reluctant to do a PR because I couldn't successfully test

eupn commented:
@dwaite No problem! I have a board and can test your changes as soon as you make a PR.

dwaite committed:
Linker fixes and update to newest crates. (#1)

- device.x: Correct linker errors by including `libnosys` and defining `end`
- .cargo/config: remove unneeded linker flags
- Cargo.toml: update to latest crates

erik-nymi commented:
I was looking to use this crate for a prototype. Is this branch close to being merged?
